File: typescript/dnd-character/dnd-character.ts
export class DnDCharacter {
  strength: number
  charisma: number
  wisdom: number
  intelligence: number
  constitution: number
  dexterity: number
  hitpoints: number

  constructor() {
    this.strength = DnDCharacter.generateAbilityScore()
    this.charisma = DnDCharacter.generateAbilityScore()
    this.wisdom = DnDCharacter.generateAbilityScore()
    this.intelligence = DnDCharacter.generateAbilityScore()
    this.constitution = DnDCharacter.generateAbilityScore()
    this.dexterity = DnDCharacter.generateAbilityScore()
    this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution)
  }

  public static generateAbilityScore(): number {
    return Math.ceil((Math.random() * 15)) + 3
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2)
  }
}

File: ruby/dnd-character/dnd_character.rb
class DndCharacter
  def self.modifier(score)
    (score - 10) / 2
  end

  attr_reader :strength,
              :dexterity,
              :constitution,
              :intelligence,
              :wisdom,
              :charisma,
              :hitpoints

  def initialize
    @strength = rand(3..18)
    @dexterity = rand(3..18)
    @constitution = rand(3..18)
    @intelligence = rand(3..18)
    @wisdom = rand(3..18)
    @charisma = rand(3..18)
    @hitpoints = 10 + DndCharacter.modifier(@constitution)
  end
end
